Update to `laminas/laminas-coding-standard:2.3.x`, improved types and internal API

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Does the donkey work upgrading to LCS 2.3

Changes to tests and src are in separate commits

gsteel committed:
Update Laminas Coding Standard to 2.3.0 and psalm to the most recent minor version

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Newly detected psalm errors base lined

These errors are primarily missing class, and missing file for legacy Zend purposes

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update phpcs ruleset with correct rule reference and xsd, ignoring cache file

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
qa: general, automated and primarily whitespace CS fixes

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
qa: Applies CS fixes to all test code, fixing new psalm errors and 544 others 💪

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
qa: apply CS fixes

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Baseline newly discovered psalm issues in src/

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Bump interop integration tests to fix test failures on lowest

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
@Ocramius …
Integration tests are failing on 8.0 and 8.1 at https://github.com/laminas/laminas-diactoros/blob/c1a61118c23a8b996082e51a3c31a07f7f7107a5/src/Stream.php#L339

I think this is because an empty path is throwing a ValueError in `fopen` - In order to fix this failure, we'll need to explicitly check for the empty path and throw the expected exception right?

gsteel commented:
OK, I take that last comment back - for some reason, the error handler is being ignored on 8.0 and 8.1 - this is all stuff that hasn't been touched by this PR so I'm a little bit confused as to why CI would be failing here…

Ocramius reviewed (approved):
Awesome, thanks @gsteel!

Ocramius commented:
Ah, I did not notice the failures. I have another branch where I did this, and will propose a patch :thinking:

Ocramius commented:
Opened #104 - unsure if failures are comparable

gsteel committed:
Removes the GMP case from the data provider

The test is checking for exceptional situations for non-stream resources and given the other cases are more likely to run (i.e. GMP is probably not installed), it's not worth the psalm errors

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Bump PHPUnit to ^9.5

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Change set/restore error handler to a try/catch block

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/Stream.php:
```diff
@@ -320,23 +318,14 @@ public function getMetadata($key = null)
+            try {
+                $resource = fopen($stream, $mode);
+            } catch (Throwable $e) {
+                throw new Exception\RuntimeException('Invalid stream reference provided');
```
We should really rethrow the previous exception

Ocramius replied:
s/rethrow/include

gsteel replied:
Do you think it's a reasonable change - tests are now passing locally for me on 8.x
… The previous behaviour was to throw what's there - I'll amend to add previous perhaps? The integration tests would fail with anything other than a RuntimeException afaict

Ocramius replied:
I think this is really only in case of a stream error - an exception should be rare, but is a healthy approach to a failure.

Ocramius reviewed (approved):
LGTM 🚢
